pkg/yurthub/cachemanager/handler_layer.go

```diff
@@ -0,0 +1,117 @@
+package cachemanager
+
+import (
+	"sync"
+
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/storage"
+	"k8s.io/klog"
+)
+
+type Handler interface {
+	ServeObject(runtime.Object) (runtime.Object, error)
+}
+
+type HandlerLayer struct {
+	sync.Mutex
+	handlers  map[schema.GroupVersionKind][]Handler
+	selectors map[schema.GroupVersionKind]*storage.SelectionPredicate
+}
+
+func NewHandlerLayer() *HandlerLayer {
+	return &HandlerLayer{
+		handlers:  make(map[schema.GroupVersionKind][]Handler),
+		selectors: make(map[schema.GroupVersionKind]*storage.SelectionPredicate),
+	}
+}
+
+//RegisterHandler is used to register the handlers of gvk, one gvk can have multiple handlers
+func (hl *HandlerLayer) RegisterHandler(gvk schema.GroupVersionKind, handler Handler) error {
+	hl.Lock()
+	defer hl.Unlock()
+	if _, ok := hl.handlers[gvk]; !ok {
+		hl.handlers[gvk] = make([]Handler, 1)
+	}
+	hl.handlers[gvk] = append(hl.handlers[gvk], handler)
+	return nil
+}
+
+//GetHandlers return the handlers according to the gvk information
+func (hl *HandlerLayer) GetHandlers(gvk schema.GroupVersionKind) []Handler {
+	if _, ok := hl.handlers[gvk]; ok {
+		return hl.handlers[gvk]
+	} else {
+		return nil
+	}
+}
+
+//RegisterSelectors is used to register the selectors of gvk, one gvk can only have one selector
+func (hl *HandlerLayer) RegisterSelectors(gvk schema.GroupVersionKind, labelSelector, fieldSelector string, parseFunc storage.AttrFunc) error {
+	hl.Lock()
+	defer hl.Unlock()
+	if _, ok := hl.selectors[gvk]; !ok {
+		parsedLabel, err := labels.Parse(labelSelector)
+		if err != nil {
+			klog.Errorf("failed to parse labels, %v", err)
+			return err
+		}
+		parsedField, err := fields.ParseSelector(fieldSelector)
+		if err != nil {
+			klog.Errorf("failed to parse fields, %v", err)
+			return err
+		}
+		hl.selectors[gvk] = &storage.SelectionPredicate{
+			Label:    parsedLabel,
+			Field:    parsedField,
+			GetAttrs: parseFunc,
+		}
+	} else {
+		klog.Errorf("the selector of %s is already exists", gvk.String())
+	}
+	return nil
+}
+
+//GetSelector return the selectors according to the gvk information
+func (hl *HandlerLayer) GetSelector(gvk schema.GroupVersionKind) *storage.SelectionPredicate {
+	if _, ok := hl.selectors[gvk]; ok {
+		return hl.selectors[gvk]
+	} else {
+		return nil
+	}
+}
+
+//Process uses the registered handlers to handle the objects, the obj passed into function shouldn't be changed
+func (hl *HandlerLayer) Process(obj runtime.Object) (runtime.Object, error) {
+	gvk := obj.GetObjectKind().GroupVersionKind()
+	handlers := hl.GetHandlers(gvk)
+	if handlers != nil {
+		var newobj runtime.Object
+		newobj = obj
```

Comment:
newobj is redundant, how about use obj directly?

Reply:
Code is already updated, thanks for review